### ISSUE TITLE
Experiment: Front-End Testing

### ISSUE THREAD

**Issue by DFurnes** [DoSomething/phoenix-next#1437]
**tl;dr:** This is a hefty write-up. If you're just interested in the takeaways, I'd recommend skimming the introduction & goals sections and then jumping right down to "recommendation" section at the end.

------------------------------------------------------------------

Phoenix currently uses [Jest](https://jestjs.io) for front-end unit & snapshot testing, covering 22 of about 100 components. Unfortunately, we've not kept pace with tests as we've added more and more functionality to this application & the "time" cost of writing the remaining tests is high.

Further, these unit tests actually don't cover large portions of our flow (such as things that happen in Redux actions or reducers), which means we still rely on manual testing and monitoring, like [Ghost Inspector](https://ghostinspector.com), to find issues.

## Goals

Our "high level" goal is to be able to rapidly iterate & deploy changes to Phoenix without fearing that we'll introduce bugs. Since we don't want to put too much feature work on hold while we write tests, we also want to make sure we get the most "bang for our buck" for the time we invest.

With that in mind, I think our time is best spent adding a core suite of _functional tests_ that run through our core flows (such as signing up & reporting back) as a user would. In particular:

- We want to test **whole website flows**, not just functions or components in isolation.
- We need to be able to **mock any state or functionality in other apps or third-party services** (e.g. Northstar, Rogue or Contentful), so that issues in other systems don't cause these tests to fail. If the tests aren't rock solid, it's tempting to just ignore them when they fail.
- We want it to be **easy and quick to run tests** during development & on a CI service (like Wercker or CircleCI) so that we limit the chance of merging bugs into `master`.

This isn't our first attempt at writing a functional test suite. We'd previously written tests with [Behat](http://behat.org/en/latest/) and [CasperJS](http://casperjs.org) for Ashes, and [Laravel Dusk](https://laravel.com/docs/5.8/dusk) for Rogue. These all suffered from the inability to mock API calls to external services & general flakiness which eroded developer trust.

We want to avoid repeating that scenario!

## Process

The landscape for front-end testing has changed pretty significantly in the past few years, so I looked at a wide array of tools as part of the "R&D" phase of this work. We got recommendations for [Cypress](https://www.cypress.io) from the folks at Artsy & [WebDriverIO](https://webdriver.io) from the testing team at the New York Times. I also looked at [Puppeteer](https://developers.google.com/web/tools/puppeteer/)/[Polly.js](https://netflix.github.io/pollyjs/#/), [jest-environment-webdriver](https://github.com/alexeyraspopov/jest-webdriver), [Nightwatch](https://nightwatchjs.org), and [Nightmare](http://www.nightmarejs.org).

I configured each of these tools on throwaway experimental branches and wrote as much of a test suite as I could to try to get a feel for how each tool worked & any notable limitations.

### Cypress

[Cypress](https://www.cypress.io) is a "complete end-to-end testing experience", with an opinionated & polished test runner, built-in assertion & mocking library, and a [SaaS product](https://www.cypress.io/dashboard/) for visualizing and sharing CI test runs. It seems to be [actively developed](https://github.com/cypress-io/cypress/graphs/contributors) by multiple "core" contributors.

I configured [an experimental branch](https://github.com/DoSomething/phoenix-next/pull/1438) in Phoenix with a test suite covering the core "signup flow" for a traditional campaign. Cypress had the most polished setup process & documentation of any of the tools I evaluated, and made an excellent first impression.

![example-cypress-cropped](https://user-images.githubusercontent.com/583202/58349124-72bc8c00-7e30-11e9-8f75-aeac88208574.gif)

They also had the best test writing experience, with a beautiful test-runner & elegant concept of ["smart retries"](https://dfurn.es/2JOB2SD) (so no more `.wait(3000)` everywhere)! It was really satisfying to write out tests & watch the browser automatically reload and run them in a second window. Having the persistent test runner was also really handy for writing & debugging tests, since you could easily "dive in" to the page and see exactly what was failing (including the ability to "time travel" to previous steps).

Unlike many of the WebDriver or Selenium-based tools I evaluated, they also support [mocking and spying](https://docs.cypress.io/guides/guides/stubs-spies-and-clocks.html) on network requests or mocking the browser's "clock" to test time-based logic. This is incredibly important for making efficient and non-flaky tests.

Their tooling integration is also the most mature of any of the tools I tested, with an [ESLint plugin](https://github.com/cypress-io/eslint-plugin-cypress), [CircleCI orb](https://github.com/cypress-io/cypress-example-circleci-orb), and TypeScript definitions for Intellisense.

I did run into a few stumbling blocks, but was able to work my way around them:

- Unfortunately, Cypress has [severe limitations on cross-domain flows](https://git.io/fj8sM) (for example, starting at `phoenix.test` and heading to `identity-dev.dosomething.org`). This made our OAuth login flow (which requires a redirect to Northstar) problematic. Luckily, I was able to instead "capture" authentication redirects & stub out `window.AUTH` object to "mock" an authenticated session without any reliance on a real authentication backend.
- Cypress doesn't support mocking `fetch` calls out-of-the-box yet (which we use for all client-side API calls), and so we have to "force" a fallback to a polyfill built on `XmlHttpRequest`. However, full network-layer stubbing is on [their roadmap](https://docs.cypress.io/guides/references/roadmap.html#Upcoming-features).
- Cypress uses Mocha's assertion and mocking syntax by default, which means some context-switching when writing different types of tests. However, it's also on [their roadmap](https://docs.cypress.io/guides/references/roadmap.html#Upcoming-features) to swap to [use Jest's assertion API](https://git.io/fj8sQ), and there's a [plugin in the meantime](https://git.io/fj8s7).
- Finally, unlike WebDriver/Selenium, Cypress only supports Chrome at the moment. Support for testing with additional browsers is [on their roadmap](https://docs.cypress.io/guides/references/roadmap.html#Upcoming-features), and is luckily not a big pain point in our current development workflow.

_Here's my [experimental branch](https://github.com/DoSomething/phoenix-next/pull/1438) with Cypress & a signup flow test suite._

### Jest, Puppeteer & Polly.js

[Puppeteer](https://developers.google.com/web/tools/puppeteer/) is a library that controls Google Chrome over the DevTools protocol, and [Polly.js](https://netflix.github.io/pollyjs/#/) is a libary that can be used to record, replay, and stub HTTP requests. This is _incredibly cool_, since it'd mean we wouldn't need to write mocks ourselves - we could just "record" API requests & responses for the desired state, and then replay those same responses in the future.

The biggest downside for these tools is that we're piecing them together ourselves & are therefore adopting more of the cost of maintaining that setup in the future. (And setup was largely a process of trial and error, since these tools are not inherently built to work together).

There's also some fun bugs, like this [known issue](https://git.io/fjRA9) where blank Chromium instances are left around after each test run (unless in headless mode). That eventually adds up:

![Screen Shot 2019-05-28 at 2 15 17 PM](https://user-images.githubusercontent.com/583202/58502227-2933b000-8154-11e9-9028-d578df3b1577.png)

I think we could likely work around these issues given time, but it's hard to argue that's a good use of resources when Cypress gives us so much of the same benefits out of the box.

_Here's my [experimental branch](https://github.com/DoSomething/phoenix-next/pull/1441) with Jest, Puppeteer, and Polly & a partially written test._

### WebDriverIO

[WebDriverIO](https://webdriver.io) is a so-called "next generation" test framework built on the [WebDriver spec](https://developer.mozilla.org/en-US/docs/Web/WebDriver). It seems to be [actively developed](https://github.com/webdriverio/webdriverio/graphs/contributors), primarily by an engineer from [Sauce Labs](https://www.saucelabs.com).

![webdriverio](https://user-images.githubusercontent.com/583202/58350363-a9e06c80-7e33-11e9-8a52-93e9f3416179.gif)

Unfortunately, setup was painful and the documentation is not very clear or consistent. Due to WebDriver's limited API, it also doesn't support mocking requests within the page (and so we'd be at the mercy of potentially brittle external dependencies).

_Here's my [experimental branch](https://github.com/DoSomething/phoenix-next/pull/1439) with WebDriverIO & a bare-bones example test._

### jest-environment-webdriver

[jest-environment-webdriver](https://github.com/alexeyraspopov/jest-webdriver) is a wrapper around Selenium's WebDriver implementation (no relation to WebDriverIO). I was really excited about this one, since it'd let us write functional tests within our existing testing framework & didn't require glueing together multiple discrete tools (like Puppeteer and Polly, above).

Unfortunately, the documentation is somewhere between sparse and non-existent & there's been [little activity since August of last year](https://github.com/alexeyraspopov/jest-webdriver/graphs/contributors). And like the other WebDriver-based tools, it has no built-in support for mocking client-side HTTP requests. I was also unable to figure out how to "await" a page load within a test, which significantly limits the scenarios we could test.

_Here's my [experimental branch](https://github.com/DoSomething/phoenix-next/compare/experimental-jest-webdriver) with jest-environment-webdriver & a bare-bones example test._

### Nightwatch

[Nightwatch](https://nightwatchjs.org) is another WebDriver-based automated testing framework. It has a relatively clean API & assertion syntax, based on [Chai](https://www.chaijs.com/api/bdd/). Like other WebDriver-based tools, it didn't have any way to mock client-side HTTP requests.

_Here's my [experimental branch](https://github.com/DoSomething/phoenix-next/compare/experimental-nightwatch) with Nightwatch & a bare-bones example test._

### Nightmare

[Nightmare](http://www.nightmarejs.org) is a high-level browser automation library from the folks at [Segment](https://segment.com). It has an elegant and simple API, and a rather nifty [Chrome extension](https://github.com/segmentio/daydream) that can be used to automatically generate scripts (for both Nightmare and Puppeteer).

Although it [seems possible to mock HTTP requests](https://github.com/segmentio/nightmare/issues/784), there's no "official" support built-in. It also doesn't seem to be [actively developed](https://github.com/segmentio/nightmare/graphs/contributors) aside from security fixes.

_I didn't end up making an experimental branch for Nightmare._

## Recommendation

Cypress was by far the most promising tool I looked at. I didn't run into nearly as many roadblocks as with the other tools I evaluated and, once I got a hang of things, the test-runner made it genuinely pleasant to write and run tests.

Whatever tool we decide to use, we'll want to look into the following:
1. Get things running on pull requests! This could be Wercker, but it might be a great opportunity to move Phoenix's CI environment [to CircleCI or another tool](https://github.com/DoSomething/infrastructure/issues/139).
2. Create some bare-bones Contentful entities (e.g. a simpler "test" campaign, etc.) and, ideally, mock them by overriding `window.STATE` when running the test suite.
3. Write more tests!!

In my mind, a good round of initial tests to prioritize include:
 - [x] standard campaign signup flow
 - [ ] standard campaign reportback flow (share, text, photo)
 - [ ] static content rendering (on a representative example page)
 - [ ] story page (especially anonymous actions)
 - [ ] homepage & explore campaigns

Finally, I focused on browser testing because it provides the greatest "bang for our buck", but we could also make improvements to component unit testing – such as swapping Enzyme to the less fiddly-looking [`react-testing-library`](https://github.com/testing-library/react-testing-library) and making more use of Apollo's [testing helpers](https://www.apollographql.com/docs/react/recipes/testing).

**Comment by DFurnes** [comment]
Going to close this out since we've settled on Cypress & started writing tests!
